File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/coroutines/async/java/UserApiClient.java
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.coroutines.async.java;

// UserApiClient simulates an external API client.
public class UserApiClient {

    public User fetchUserById(String userId) {
        // This method blocks the current thread while waiting for the response.
        return new User(userId, "John Doe");
    }
}

File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/coroutines/async/kotlin/UserApiClient.kt
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.coroutines.async.kotlin

// UserApiClient simulates an external API client.
class UserApiClient {

    suspend fun fetchUserById(userId: String): User {
        // suspend means this function can pause without blocking a thread.
        return User(userId, "John Doe")
    }
}